MaterialDesignThemes.Wpf/NumericUpDown.cs

```diff
@@ -0,0 +1,257 @@
+    }
+    #endregion Event : ValueChangedEvent
+
+    public override void OnApplyTemplate()
+    {
+        if (_increaseButton != null)
+            _increaseButton.Click -= IncreaseButtonOnClick;
+
+        if (_decreaseButton != null)
+            _decreaseButton.Click -= DecreaseButtonOnClick;
+        if (_textBoxField != null)
+            _textBoxField.TextChanged -= OnTextBoxFocusLost;
+
+        _increaseButton = GetTemplateChild(IncreaseButtonPartName) as RepeatButton;
+        _decreaseButton = GetTemplateChild(DecreaseButtonPartName) as RepeatButton;
+        _textBoxField = GetTemplateChild(TextFieldBoxPartName) as TextBox;
+
+        if (_increaseButton != null)
+            _increaseButton.Click += IncreaseButtonOnClick;
+
+        if (_decreaseButton != null)
+            _decreaseButton.Click += DecreaseButtonOnClick;
+
+        if (_textBoxField != null)
+        {
+            _textBoxField.LostFocus += OnTextBoxFocusLost;
+            _textBoxField.Text = Value.ToString(CultureInfo.CurrentUICulture);
+        }
+
+        base.OnApplyTemplate();
+    }
+
+    private void OnTextBoxFocusLost(object sender, EventArgs e)
+    {
+        if (_textBoxField is { } textBoxField)
+        {
+            if (int.TryParse(textBoxField.Text, NumberStyles.Integer, CultureInfo.CurrentUICulture, out int numericValue))
+            {
+                SetCurrentValue(ValueProperty, numericValue);
+            }
+            else
+            {
+                textBoxField.Text = Value.ToString(CultureInfo.CurrentUICulture);
+            }
+        }
+    }
+
+    private void IncreaseButtonOnClick(object sender, RoutedEventArgs e) => OnIncrease();
+
+    private void DecreaseButtonOnClick(object sender, RoutedEventArgs e) => OnDecrease();
+
+    private void OnIncrease()
+    {
+        SetCurrentValue(ValueProperty, Value + 1);
+
+        if (IncreaseCommand?.CanExecute(IncreaseCommandParameter) ?? false)
+            IncreaseCommand.Execute(IncreaseCommandParameter);
+    }
+
+    private void OnDecrease()
+    {
+        SetCurrentValue(ValueProperty, Value - 1);
+
+        if (DecreaseCommand?.CanExecute(this) ?? false)
+            DecreaseCommand.Execute(this);
+    }
+
+    protected override void OnPreviewKeyDown(KeyEventArgs e)
+    {
+        if (e.Key == Key.Up)
+        {
+            OnIncrease();
+            e.Handled = true;
+        }
+        else if (e.Key == Key.Down)
+        {
+            OnDecrease();
+            e.Handled = true;
+        }
+        base.OnPreviewKeyDown(e);
+    }
+
+    protected override void OnPreviewMouseWheel(MouseWheelEventArgs e)
+    {
+        if (IsKeyboardFocusWithin && AllowChangeOnScroll)
+        {
+            if (e.Delta > 0)
+            {
+                OnIncrease();
+                e.Handled = true;
+            }
+            else if (e.Delta < 0)
+            {
+                OnDecrease();
+                e.Handled = true;
+            }
```

Comment:
Not sure if it is an actual issue, but this will not mark the event as handled if `e.Delta == 0` which seems a bit strange.

Imagine if the control is located in a `ScrollViewer`, it has keyboard focus you the user user the scroll wheel to change the values. Would there be a scenario where you think you are scrolling the values, but hit the edge case and suddenly the `ScrollViewer` scrolls instead because the event bubbles up? That would be annoying!